src/rust/engine/stdio/src/lib.rs

```diff
@@ -0,0 +1,439 @@
+// Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+#![deny(warnings)]
+// Enable all clippy lints except for many of the pedantic ones. It's a shame this needs to be copied and pasted across crates, but there doesn't appear to be a way to include inner attributes from a common source.
+#![deny(
+  clippy::all,
+  clippy::default_trait_access,
+  clippy::expl_impl_clone_on_copy,
+  clippy::if_not_else,
+  clippy::needless_continue,
+  clippy::unseparated_literal_suffix,
+  // TODO: Falsely triggers for async/await:
+  //   see https://github.com/rust-lang/rust-clippy/issues/5360
+  // clippy::used_underscore_binding
+)]
+// It is often more clear to show that nothing is being moved.
+#![allow(clippy::match_ref_pats)]
+// Subjective style.
+#![allow(
+  clippy::len_without_is_empty,
+  clippy::redundant_field_names,
+  clippy::too_many_arguments
+)]
+// Default isn't as big a deal as people seem to think it is.
+#![allow(clippy::new_without_default, clippy::new_ret_no_self)]
+// Arc<Mutex> can be more clear than needing to grok Orderings:
+#![allow(clippy::mutex_atomic)]
+
+mod term;
+
+pub use term::{TermReadDestination, TermWriteDestination, TryCloneAsFile};
+
+use std::cell::RefCell;
+use std::fmt;
+use std::fs::File;
+use std::future::Future;
+use std::io::{Read, Write};
+use std::os::unix::io::{AsRawFd, FromRawFd, IntoRawFd, RawFd};
+use std::sync::Arc;
+
+use parking_lot::Mutex;
+use tokio::task_local;
+
+///
+/// A Console wraps some "borrowed" file handles: when it is dropped, we forget about the file
+/// handles rather than closing them. The file handles are optional only so that they may be
+/// "taken" during Drop.
+///
+#[derive(Debug)]
+struct Console {
+  stdin_handle: Option<File>,
+  stdout_handle: Option<File>,
+  stderr_handle: Option<File>,
+}
+
+impl Console {
+  fn new(stdin_fd: RawFd, stdout_fd: RawFd, stderr_fd: RawFd) -> Console {
+    let (stdin, stdout, stderr) = unsafe {
+      (
+        File::from_raw_fd(stdin_fd),
+        File::from_raw_fd(stdout_fd),
+        File::from_raw_fd(stderr_fd),
+      )
+    };
```

Comment:
I haven't worked with unsafe before. If this fails, will we get a panic?

Reply:
In the majority of cases, unsafe wraps "memory safety" cases... in those cases, using unsafe can cause segfaults (which are generally worse than panics, because you don't get a backtrace).

In _this_ case, the unsafety is around file handles, and generally deals with files being closed early or multiple times. Rather than memory safety, that will "just" cause operations on those file handles to fail. But since any meaningful operation on a file handle involves a syscall, they are all already fallible.

The case that could happen here is that we could try to open a file handle for a "bogus" file. But the consequence of that is an IO error the first time you deal with it ("file is closed", etc).

I'll add a comment nearby the other usage of unsafe in here that _actually_ caused me some pain =)